lib/api/unittest/CDataFrameAnalyzerTrainingTest.cc

```diff
@@ -817,4 +817,128 @@ BOOST_AUTO_TEST_CASE(testCategoricalFieldsEmptyAsMissing) {
     });
 }
 
+BOOST_AUTO_TEST_CASE(testRunBoostedTreeRegressionFeatureImportance) {
+
+    // Test that feature importance correctly recognize the impact of regressors in a linear model.
+
+    double alpha{2.0};
+    double lambda{1.0};
+    double gamma{10.0};
+    double softTreeDepthLimit{5.0};
+    double softTreeDepthTolerance{0.1};
+    double eta{0.9};
+    std::size_t maximumNumberTrees{1};
+    double featureBagFraction{1.0};
+
+    int rows = 200;
+
+    // test with all  shap values
+    {
+        std::stringstream output;
+        auto outputWriterFactory = [&output]() {
+            return std::make_unique<core::CJsonOutputStreamWrapper>(output);
+        };
+
+        api::CDataFrameAnalyzer analyzer{
+            test::CDataFrameAnalysisSpecificationFactory::predictionSpec(
+                "regression", "c5", rows, 5, 4000000, 0, 0, {"c1"}, alpha,
+                lambda, gamma, softTreeDepthLimit, softTreeDepthTolerance, eta,
+                maximumNumberTrees, featureBagFraction, 4),
+            outputWriterFactory};
+
+        TDoubleVec expectedPredictions;
+
+        TStrVec fieldNames{"c1", "c2", "c3", "c4", "c5", ".", "."};
+        TStrVec fieldValues{"", "", "", "", "", "0", ""};
+        test::CRandomNumbers rng;
+        TDoubleVec weights{50, 150, 50, -50};
+
+        TDoubleVec values;
+        rng.generateUniformSamples(-10.0, 10.0, weights.size() * rows, values);
+
+        // make last column categorical
+        for (auto it = values.begin(); it < values.end(); it += 4) {
+            *it = (*it < 0) ? -10 : 10;
+        }
+
+        auto frame = setupLinearRegressionData(fieldNames, fieldValues,
+                                               analyzer, weights, values);
+        analyzer.handleRecord(fieldNames, {"", "", "", "", "", "", "$"});
+
+        rapidjson::Document results;
+        rapidjson::ParseResult ok(results.Parse(output.str()));
+        BOOST_TEST_REQUIRE(static_cast<bool>(ok) == true);
+
+        double c1_sum, c2_sum, c3_sum, c4_sum;
+        for (const auto& result : results.GetArray()) {
+            if (result.HasMember("row_results")) {
+                c1_sum += std::fabs(
+                    result["row_results"]["results"]["ml"]["shap.c1"].GetDouble());
+                c2_sum += std::fabs(
+                    result["row_results"]["results"]["ml"]["shap.c2"].GetDouble());
+                c3_sum += std::fabs(
+                    result["row_results"]["results"]["ml"]["shap.c3"].GetDouble());
+                c4_sum += std::fabs(
+                    result["row_results"]["results"]["ml"]["shap.c4"].GetDouble());
+            }
+        }
```

Comment:
I think it might be useful to state the linear function we're fitting here to explain these inequalities, i.e. `50 c_1 + 150 c_2 + 50 c_3 - 50 c_4`. This leads to a question: why do we expect `c1_sum` > `c3_sum`, since the feature values come from the same distribution? Isn't this just an artefact of sampling, in which case could we maybe make w_3 < w_1 < w_2.

Reply:
Ok, reading the code above more carefully, I think this comes from the fact that mean(|c1 feature values|) > mean(|c3 feature values|), etc. So I think the test is good as is, but this needs more explanation.

Reply:
Doc adjusted.